File: packages/depz-sensor-sdk-java/src/main/java/ai/depz/sensor/protocol/FwDepz.java
package ai.depz.sensor.protocol;

import java.util.Arrays;

import ai.depz.sensor.transport.Crc;

/** {@code .fwdepz} bootloader container parse/validate (contracts/06 §2). */
public final class FwDepz {
    private FwDepz() {}

    public static final byte[] FWDEPZ_MAGIC = "FWDEPZ00".getBytes(java.nio.charset.StandardCharsets.US_ASCII);
    public static final int FWDEPZ_HEADER_SIZE = 64;

    /** Reason codes match the vector {@code error} strings. */
    public static final class FwDepzError extends RuntimeException {
        public final String code;
        public FwDepzError(String code, String message) {
            super(message);
            this.code = code;
        }
    }

    /** Parsed and validated {@code .fwdepz} firmware container. */
    public record FwDepzImage(long loadAddr, long fwSize, long fwCrc32, int curSec, int totSec, byte[] payload) {

        public boolean payloadCrcOk() {
            return Crc.crc32IsoHdlc(payload) == fwCrc32;
        }

        /**
         * Parse and validate. Validation order (contract 06 §2): length, magic,
         * then CRC-16/CCITT-FALSE over bytes [0..61] vs the u16 LE at offset 62,
         * then {@code fw_size == payload length}.
         */
        public static FwDepzImage parse(byte[] blob) {
            if (blob.length < FWDEPZ_HEADER_SIZE) {
                throw new FwDepzError("too_short", "file too short: " + blob.length + " < " + FWDEPZ_HEADER_SIZE);
            }
            if (!Arrays.equals(Arrays.copyOfRange(blob, 0, 8), FWDEPZ_MAGIC)) {
                throw new FwDepzError("magic", "bad magic (not a .fwdepz file)");
            }
            int hdrCrc = (blob[62] & 0xFF) | ((blob[63] & 0xFF) << 8);
            int actual = Crc.crc16CcittFalse(Arrays.copyOfRange(blob, 0, 62));
            if (hdrCrc != actual) {
                throw new FwDepzError("header_crc",
                    String.format("header CRC mismatch: stored=0x%04X actual=0x%04X", hdrCrc, actual));
            }
            long loadAddr = u32le(blob, 8);
            long fwSize = u32le(blob, 12);
            long fwCrc32 = u32le(blob, 16);
            int curSec = blob[20] & 0xFF;
            int totSec = blob[21] & 0xFF;
            byte[] payload = Arrays.copyOfRange(blob, FWDEPZ_HEADER_SIZE, blob.length);
            if (fwSize != payload.length) {
                throw new FwDepzError("size", "fw_size=" + fwSize + " but payload is " + payload.length + " bytes");
            }
            return new FwDepzImage(loadAddr, fwSize, fwCrc32, curSec, totSec, payload);
        }
    }

    private static long u32le(byte[] b, int off) {
        return (b[off] & 0xFFL) | ((b[off + 1] & 0xFFL) << 8) | ((b[off + 2] & 0xFFL) << 16) | ((b[off + 3] & 0xFFL) << 24);
    }
}

File: packages/depz-sensor-sdk-csharp/src/Depz.Sensor/Protocol/FwDepz.cs
using System.Buffers.Binary;
using System.Text;
using Depz.Sensor.Transport;

namespace Depz.Sensor.Protocol;

/// <summary><c>.fwdepz</c> bootloader container parse/validate (contracts/06 §2).</summary>
public static class FwDepz
{
    public static readonly byte[] FwDepzMagic = Encoding.ASCII.GetBytes("FWDEPZ00");
    public const int FwDepzHeaderSize = 64;

    /// <summary>Parse/validate failure. <see cref="Code"/> matches the vector <c>error</c> strings.</summary>
    public sealed class FwDepzException : Exception
    {
        public string Code { get; }
        public FwDepzException(string code, string message) : base(message) => Code = code;
    }

    /// <summary>Parsed and validated <c>.fwdepz</c> firmware container.</summary>
    public sealed record FwDepzImage(uint LoadAddr, uint FwSize, uint FwCrc32, int CurSec, int TotSec, byte[] Payload)
    {
        public bool PayloadCrcOk => Crc.Crc32IsoHdlc(Payload) == FwCrc32;

        /// <summary>
        /// Parse and validate. Validation order (contract 06 §2): length, magic,
        /// then CRC-16/CCITT-FALSE over bytes [0..61] vs the u16 LE at offset 62,
        /// then fw_size == payload length.
        /// </summary>
        public static FwDepzImage Parse(byte[] blob)
        {
            if (blob.Length < FwDepzHeaderSize)
                throw new FwDepzException("too_short", $"file too short: {blob.Length} < {FwDepzHeaderSize}");

            if (!blob.AsSpan(0, 8).SequenceEqual(FwDepzMagic))
                throw new FwDepzException("magic", "bad magic (not a .fwdepz file)");

            int hdrCrc = blob[62] | (blob[63] << 8);
            ushort actual = Crc.Crc16CcittFalse(blob.AsSpan(0, 62));
            if (hdrCrc != actual)
                throw new FwDepzException("header_crc",
                    $"header CRC mismatch: stored=0x{hdrCrc:X4} actual=0x{actual:X4}");

            uint loadAddr = BinaryPrimitives.ReadUInt32LittleEndian(blob.AsSpan(8));
            uint fwSize = BinaryPrimitives.ReadUInt32LittleEndian(blob.AsSpan(12));
            uint fwCrc32 = BinaryPrimitives.ReadUInt32LittleEndian(blob.AsSpan(16));
            int curSec = blob[20];
            int totSec = blob[21];
            byte[] payload = blob[FwDepzHeaderSize..];

            if (fwSize != payload.Length)
                throw new FwDepzException("size", $"fw_size={fwSize} but payload is {payload.Length} bytes");

            return new FwDepzImage(loadAddr, fwSize, fwCrc32, curSec, totSec, payload);
        }
    }
}
